api/envoy/config/route/v3/route_components.proto

```diff
@@ -1593,3 +1600,30 @@ message QueryParameterMatcher {
     bool present_match = 6;
   }
 }
+
+// HTTP Internal Redirect :ref:`architecture overview <arch_overview_internal_redirects>`.
+message InternalRedirectPolicy {
+  // An internal redirect is not handled, unless the number of previous internal redirects that a
+  // downstream request has encountered is lower than this value.
+  // In the case where a downstream request is bounced among multiple routes by internal redirect,
+  // the first route that hits this threshold, or does not set :ref:`internal_redirect_policy
+  // <envoy_api_field_config.route.v3.RouteAction.internal_redirect_policy>`
+  // will pass the redirect back to downstream.
+  //
+  // If not specified, at most one redirect will be followed.
+  google.protobuf.UInt32Value max_internal_redirects = 1;
+
+  // Defines what upstream response codes are allowed to trigger internal redirect. If unspecified,
+  // only 302 will be treated as internal redirect.
+  // Only 301, 302, 303, 307 and 308 are valid values. Any other codes will be ignored.
+  repeated uint32 redirect_response_codes = 2 [(validate.rules).repeated = {max_items: 5}];
+
+  // Specifies a list of predicates that are queried when an upstream response is deemed
+  // to trigger an internal redirect by all other criteria. Any predicate in the list can reject
+  // the redirect, causing the response to be proxied to downstream.
+  repeated google.protobuf.Any predicates = 3;
+
+  // Disallow internal redirect to follow a target URI with a different scheme than the value of
+  // x-forwarded-proto. Default behavior is to follow such redirects.
```

Comment:
Yeah. This is a good point. While I think writing the OnlyAllowSafeCrossSchemeRedirect predicate definitely is a good way to balance API simplicity and safety, I think there is an alternative to it by renaming disallow_cross_scheme_redirect (yet again) to allow_only_safer_cross_scheme_redirect?

I think the advantage of this renaming is that: we keep the default behavior if nothing is said without needing to additionally config a predicate. This is slightly complicated API, since we're pushing the balance a little more to the safety side, but we do get a bit of simplicity in terms of default behavior.

I cannot decide which approach I like better. What do you think?

Reply:
Added the OnlyAllowSafeCrossSchemeRedirect.

Reply:
@alyssawilk PTAL.

@htuch PTAL again since I made some API change.

Reply:
Yep. I think I actually like making the new default being stricter - it's matching the previous documented behavior. With the predicate we'll be restore the implemented behavior. Done :)